### PR TITLE
Source build of curl to enable proxy support for Pkg

### LINKED ISSUE #13472: libcurl problems with libgit2 on master
I'm seeing the following issue when trying to use `Pkg.init()` on a fresh build of master:

```
$ ./julia 
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.5.0-dev+640 (2015-10-06 18:24 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 128f8c6 (0 days old master)
|__/                   |  x86_64-linux-gnu

julia> Pkg.init()
INFO: Initializing package repository /home/lkuper/.julia/v0.5
INFO: Cloning METADATA from https://github.com/JuliaLang/METADATA.jl
ERROR: GitError(Code:ERROR, Class:Net, SSL error: error:140E0114:SSL routines:SSL_shutdown:uninitialized)
 [inlined code] from libgit2/error.jl:96
 in clone at libgit2/repository.jl:185
 in clone at libgit2.jl:303
 in anonymous at pkg/dir.jl:49
 in cd at file.jl:22
 in init at pkg/dir.jl:47
 in init at pkg.jl:23
```

I started seeing this after pulling and rebuilding to fix the libgit2 SSL certificate issue in #13399.  I'm now seeing this error instead of the certificate error.  There's more discussion on that bug.

This is on Ubuntu 14.04, and I have `libssl-dev` installed.  I've already tried `make -C deps distclean-libgit2`.  I even did a `make distcleanall` to make extra sure.  

Before I deleted my `~/.julia/v0.5` directory (because I hoped that doing that and running `Pkg.init()` might solve the problem), I was seeing a similar error message from `Pkg.update()`.

I'm also having the same problem on CentOS with `openssl-devel` installed.

### PR DESCRIPTION
**Pull request by ViralBShah**
Using libcurl with libgit2 will allow the use of proxy servers for those behind firewalls. 

Should fix #13472 #10754

### CODE REVIEW

File: deps/curl.mk [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+$(SRCDIR)/srccache/curl-$(CURL_VER).tar.bz2: | $(SRCDIR)/srccache
+	$(JLDOWNLOAD) $@ https://curl.haxx.se/download/curl-$(CURL_VER).tar.bz2
+$(SRCDIR)/srccache/curl-$(CURL_VER)/configure: $(SRCDIR)/srccache/curl-$(CURL_VER).tar.bz2
+	$(JLCHECKSUM) $<
+	cd $(dir $<) && $(TAR) jxf $(notdir $<)
+	touch -c $@
+$(BUILDDIR)/curl-$(CURL_VER)/config.status: $(SRCDIR)/srccache/curl-$(CURL_VER)/configure
+	mkdir -p $(dir $@)
+	cd $(dir $@) && \
+	$< $(CONFIGURE_COMMON) --includedir=$(build_includedir) --with-mbedtls=$(build_shlibdir) CFLAGS="$(CFLAGS) $(CURL_CFLAGS)" LDFLAGS="$(LDFLAGS) $(CURL_LDFLAGS)"
```
Comment (tkelman) [inline deps/curl.mk]: I think you may need to also do `--disable-ssl` or something like that to get it to not link against openssl?

File: deps/patches/mbedtls-ssl.h.patch [hunk 2]
```diff
@@ -0,0 +1,12 @@
+--- ssl.h.old	2016-06-28 18:12:06.000000000 +0530
++++ ssl.h	2016-08-03 18:51:34.000000000 +0530
+@@ -54,7 +54,8 @@
+ #endif
+ 
+ #if defined(MBEDTLS_HAVE_TIME)
+-#include <time.h>
++//#include <time.h>
++#include "platform.h"
```
Comment (tkelman) [inline deps/patches/mbedtls-ssl.h.patch]: strange - cmake should be doing the configure probes for this, what were you seeing without this patch?

**Comment by tkelman** [comment]
Do we have a test system with a proxy set up somewhere that we can run this on?

File: deps/patches/mbedtls-ssl.h.patch (hunk 2 above)
Comment (ViralBShah) [inline deps/patches/mbedtls-ssl.h.patch]: This is a known bug fixed upstream, and hopefully should be in mbedtls 2.3.1. Maybe mac only, but the fix should work everywhere (I think).

https://github.com/ARMmbed/mbedtls/issues/522

**Comment by ViralBShah** [comment]
Seems like everything is building and linking ok:
````
viral-laptop 07:43:29 (vs/curl) ~/julia/usr/lib$ otool -L libgit2.dylib 
libgit2.dylib:
	@rpath/libgit2.dylib (compatibility version 24.0.0, current version 0.24.0)
	@rpath/libcurl.dylib (compatibility version 9.0.0, current version 9.0.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.5)
	/System/Library/Frameworks/Security.framework/Versions/A/Security (compatibility version 1.0.0, current version 57337.60.2)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1258.1.0)
	@rpath/libssh2.dylib (compatibility version 1.0.0, current version 1.0.1)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1226.10.1)
viral-laptop 07:43:44 (vs/curl) ~/julia/usr/lib$ otool -L libcurl.dylib 
libcurl.dylib:
	@rpath/libcurl.dylib (compatibility version 9.0.0, current version 9.0.0)
	@rpath/libssh2.dylib (compatibility version 1.0.0, current version 1.0.1)
	@rpath/libmbedtls.dylib (compatibility version 10.0.0, current version 2.3.0)
	@rpath/libmbedx509.dylib (compatibility version 0.0.0, current version 2.3.0)
	@rpath/libmbedcrypto.dylib (compatibility version 0.0.0, current version 2.3.0)
	/System/Library/Frameworks/LDAP.framework/Versions/A/LDAP (compatibility version 1.0.0, current version 2.4.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.5)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1226.10.1)
````

**Comment by ViralBShah** [comment]
@mdpradeep I believe you had a test system with proxy setup. Could you test this PR?

I also believe I can setup a SOCKS proxy using ssh -D and try. Does anyone have thoughts if this could be a good way to test?

File: deps/curl.mk (hunk 1 above)
Comment (ViralBShah) [inline deps/curl.mk]: It doesn't seem to be linking against openssl. See my `otool` output below.

**Comment by ViralBShah** [comment]
@tkelman Does this PR need anything further before asking others to try it out?

File: deps/Makefile [hunk 3]
```diff
@@ -272,6 +276,7 @@ include $(SRCDIR)/patchelf.mk
 include $(SRCDIR)/mbedtls.mk
 include $(SRCDIR)/libssh2.mk
 include $(SRCDIR)/libgit2.mk
+include $(SRCDIR)/curl.mk
```
Comment (tkelman) [inline deps/Makefile]: should be done before libgit2, and libgit2 will need to declare a dependency on it

**Comment by tkelman** [comment]
I don't think we should rush merging this into rc1, but people can certainly test the branch.

Minor thing where it'll need mentioning:
- [ ] license
- [ ] readme
- [ ] `JL_PRIVATE_LIBS`
- [ ] the `getall` target
- [ ] the list of `# autoconf configure-driven scripts: `
- [ ] tests, libgit2 features probably?

**Comment by ViralBShah** [comment]
Yes, this is definitely not for rc1, but if we get all the testing done, I hope we can merge it into rc2. If we can confirm success on all the platforms, then doing the rest of these items should be straightforward.

File: deps/libgit2.mk [hunk 4]
```diff
@@ -7,7 +7,7 @@ $(eval $(call git-external,libgit2,LIBGIT2,CMakeLists.txt,build/libgit2.$(SHLIB_
 LIBGIT2_OBJ_SOURCE := $(BUILDDIR)/$(LIBGIT2_SRC_DIR)/libgit2.$(SHLIB_EXT)
 LIBGIT2_OBJ_TARGET := $(build_shlibdir)/libgit2.$(SHLIB_EXT)
 
-LIBGIT2_OPTS := $(CMAKE_COMMON) -DCMAKE_BUILD_TYPE=Release -DTHREADSAFE=ON
+LIBGIT2_OPTS := $(CMAKE_COMMON) -DCMAKE_BUILD_TYPE=Release -DTHREADSAFE=ON -DCURL_INCLUDE_DIRS=$(build_includedir) -DCURL_LIBRARIES="-L$(build_shlibdir) -lcurl"
```
Comment (tkelman) [inline deps/libgit2.mk]: I don't think this entirely works
```
  Manually-specified variables were not used by the project:

    CMAKE_CXX_COMPILER
    CMAKE_CXX_COMPILER_ARG1
    CURL_INCLUDE_DIRS
    CURL_LIBRARIES
```

Comment (ViralBShah) [inline deps/libgit2.mk]: Is it getting auto-detected? How do I provide those variables to the build?

Comment (tkelman) [inline deps/libgit2.mk]: https://github.com/libgit2/libgit2/blob/152efee20b74ea261cf8e05410a110687e17376e/CMakeLists.txt#L277-L280 - it uses `PKG_CHECK_MODULES` which means it's relying on `pkg-config` to do the finding. I don't think we've been explicitly requiring or checking pkg-config as a build dependency though.

Comment (ViralBShah) [inline deps/libgit2.mk]: I believe pkg-config is getting used in many other places. For now, should I just remove these variables and we should check if everything works on other OSes.

Comment (tkelman) [inline deps/libgit2.mk]: almost entirely optionally though. I've had to remove it in the past to get a multilib build of 32 bit linux Julia to work from 64 bit linux

Comment (tkelman) [inline deps/libgit2.mk]: x-ref https://github.com/JuliaLang/julia/blob/master/deps/patches/libgit2-ssh.patch

**Comment by mdpradeep** [comment]
Will check it.

**Comment by wildart** [comment]
`pkg-config` should work if `libcurl` pc-file is correctly installed in `lib/pkg-config`.

File: deps/libgit2.mk (hunk 4 above)
Comment (wildart) [inline deps/libgit2.mk]: `pkg-config` should work if `libcurl` pc-file is correctly installed in `lib/pkg-config`, no need for `CURL_INCLUDE_DIRS` and `CURL_LIBRARIES`

Comment (tkelman) [inline deps/libgit2.mk]: If we need pkg-config to be present to get a correct build, then we should be checking for it and failing the build if it isn't (and documenting this). Can we get away without it?

Comment (wildart) [inline deps/libgit2.mk]: no, unless `libcurl` generates `cmake` installation configs as `libssh2` does.

**Comment by ViralBShah** [comment]
I set up a local socks proxy with `ssh -v -D localhost:2222 localhost`, and then exported `ALL_PROXY=socks5h://localhost:2222`. Starting Julia and doing `Pkg.update()` then did show connections being routed through the proxy - which suggests that this works at least on the mac.

File: deps/libgit2.mk (hunk 4 above)
Comment (ViralBShah) [inline deps/libgit2.mk]: I should mention that I do not get the unused variables during cmake that you posted above. I just get the first two but not the curl ones.

File: deps/Makefile (hunk 3 above)
Comment (ViralBShah) [inline deps/Makefile]: Good point. libgit2 and curl are in different build stages. Shouldn't that be good enough? I do think that libssh2 should probably be stage 1 (currently in 2), curl in stage 2 and libgit2 in stage 3.

**Comment by ViralBShah** [comment]
Seems like we need 4 build stages: 

libgit2, which needs libcurl, which needs libssh2, which needs mbedtls.

**Comment by wildart** [comment]
@ViralBShah You need to make sure that `libcurl` is build before `libgit2` is configured. It would not be able to use it.

**Comment by tkelman** [comment]
do we need curl linked against libssh?

**Comment by wildart** [comment]
no need

File: deps/curl.mk [hunk 5]
```diff
@@ -0,0 +1,46 @@
+## CURL ##
+
+CURL_SRC_TARGET := $(BUILDDIR)/curl-$(CURL_VER)/.libs/libcurl.$(SHLIB_EXT)
+CURL_OBJ_TARGET := $(build_shlibdir)/libcurl.$(SHLIB_EXT)
+
+ifneq ($(OS),WINNT)
+CURL_LDFLAGS := "-Wl,-rpath,'$(build_libdir)'"
```
Comment (tkelman) [inline deps/curl.mk]: I don't think this is enough, ref https://github.com/JuliaLang/julia/pull/17458/files

**Comment by ViralBShah** [comment]
How do I ensure libcurl is built before libgit2? We don't have enough build stages for all the dependencies. Ok to just go with 4 build stages? Should be straightforward.

As for linking to libssh2 - curl is doing it on its own. Is it a problem? I have no idea how to disable it if we have to.

**Comment by ViralBShah** [comment]
I also have no idea how to write tests for this that will test working through a proxy. The existing Pkg tests are sufficient otherwise.

**Comment by tkelman** [comment]
test the libgit2 feature flags to ensure it was built against curl, like is done for ssh support

as long as you set `DEP_LIBS := $(STAGE1_DEPS) $(STAGE2_DEPS) $(STAGE3_DEPS) $(STAGE4_DEPS)` and the right line in the `contrib/windows/msys_build.sh` appveyor script so it doesn't build from source every time there

File: deps/curl.mk (hunk 5 above)
Comment (ViralBShah) [inline deps/curl.mk]: I copied the pcre.mk file. Does this mean it also needs updating?

Comment (tkelman) [inline deps/curl.mk]: I think all of our rpaths may be pretty broken right now, ref https://github.com/JuliaLang/julia/issues/17602#issuecomment-237190583

**Comment by aviks** [comment]
Trying this on linux, i don't seem to get libgit2 compiled with libcurl.  when i try package operations, it does not go thru the proxy. 

```
git status
On branch vs/curl
Your branch is up-to-date with 'origin/vs/curl'.
```

```
ldd usr/lib/libgit2.so
        linux-vdso.so.1 =>  (0x00007ffe4fb23000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007faa8a2c6000)
        libssl.so.1.0.0 => /home/aviks/dev/julia5/usr/lib/libssl.so.1.0.0 (0x00007faa8a067000)
        libcrypto.so.1.0.0 => /home/aviks/dev/julia5/usr/lib/libcrypto.so.1.0.0 (0x00007faa89c8b000)
        libssh2.so.1 => /home/aviks/dev/julia5/usr/lib/libssh2.so.1 (0x00007faa89a58000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007faa89850000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007faa89632000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007faa8926d000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007faa89069000)
        libmbedcrypto.so.0 => /home/aviks/dev/julia5/usr/lib/libmbedcrypto.so.0 (0x00007faa88e16000)
        /lib64/ld-linux-x86-64.so.2 (0x00007faa8a7de000)
```

**Comment by ViralBShah** [comment]
This might be the build stages issue. I am pushing a fix shortly and hoping it gets it right. Can you manually distclean and build libcurl first and then distclean and build libgit2 afterwards? Then hopefully the libgit2 cmake should detect libcurl.

**Comment by ViralBShah** [comment]
@tkelman What is the right line to be put in mysys_build.sh? Is it:

````
echo 'override STAGE3_DEPS = ' >> Make.user
````

**Comment by ViralBShah** [comment]
@aviks I have pushed the dependency ordering to this branch.

**Comment by tkelman** [comment]
If you're introducing a `STAGE4_DEPS`, then should set it to empty on appveyor so it doesn't build from source, yes. We should probably disable building libcurl on windows anyway since libgit2 won't link to it there without more patching.

File: deps/libgit2.mk (hunk 4 above)
Comment (tkelman) [inline deps/libgit2.mk]: it's on windows - libgit2 won't use curl there if it's using winhttp

**Comment by vtjnash** [comment]
That STAGE stuff should be considered to be really old, vestigial, and possibly broken. There shouldn't be a need to add stuff there. Eventually, someone will get around to cleaning it up.

**Comment by tkelman** [comment]
actually I don't think it does anything, they just get appended into `DEP_LIBS`

**Comment by ViralBShah** [comment]
So - then how do I get build dependencies specified?

**Comment by wildart** [comment]
add `$(CURL_OBJ_TARGET)` after `$(LIBSSH2_OBJ_TARGET)` in `libgit2.mk:47`

**Comment by ViralBShah** [comment]
Ok - I just thought there may be some new way of doing things. 

I see this block in libgit2.mk - is it ok? Shouldn't the linux block have all the other dependencies as well, and the non-linux stuff be in an else block?

````
ifeq ($(OS),Linux)
$(BUILDDIR)/$(LIBGIT2_SRC_DIR)/Makefile: $(SRCDIR)/srccache/$(LIBGIT2_SRC_DIR)/libgit2-require-openssl.patch-applied $(SRCDIR)/srccache/$(LIBGIT2_SRC_DIR)/libgit2-openssl-hang.patch-applied
endif
$(BUILDDIR)/$(LIBGIT2_SRC_DIR)/Makefile: $(SRCDIR)/srccache/$(LIBGIT2_SRC_DIR)/CMakeLists.txt $(SRCDIR)/srccache/$(LIBGIT2_SRC_DIR)/libgit2-ssh.patch-applied $(SRCDIR)/srccache/$(LIBGIT2_SRC_DIR)/libgit2-agent-nonfatal.patch-applied $(LIBSSH2_OBJ_TARGET) $(CURL_OBJ_TARGET)
        mkdir -p $(dir $@)
        cd $(dir $@) && \
        $(CMAKE) $(dir $<) $(LIBGIT2_OPTS)
        touch -c $@
````

**Comment by tkelman** [comment]
you can list dependencies for the same target on multiple lines, and it will depend on all of them

**Comment by ViralBShah** [comment]
That is convenient. I didn't know that.

**Comment by ViralBShah** [comment]
On testing - `libgit2.features()` doesn't say anything about curl.

@aviks 's failure on linux was due to the absence of `pkg-config`.  I had the same issue and after apt-getting it, everything worked out fine.

**Comment by ViralBShah** [comment]
I think this is ready to merge on master. Need to decide:

1. What to do on windows? If we should not build it there, what's the best way to do so? I say that if curl gets built on windows but not used by libgit2, it's ok for now.
2. I doubt there's anything we can do on tests, unless we can query something apart from Libgit2.features(). This is all it has:
````
int git_libgit2_features()
{
        return 0
#ifdef GIT_THREADS
                | GIT_FEATURE_THREADS
#endif
#if defined(GIT_OPENSSL) || defined(GIT_WINHTTP) || defined(GIT_SECURE_TRANSPORT)
                | GIT_FEATURE_HTTPS
#endif
#if defined(GIT_SSH)
                | GIT_FEATURE_SSH
#endif
#if defined(GIT_USE_NSEC)
                | GIT_FEATURE_NSEC
#endif
        ;
}
````
3. Document `pkg-config` as a build dependency.

**Comment by tkelman** [comment]
not just document, but enforce. if we need pkg-config for this to work, then we need to enforce that rather than allowing silently broken builds if it's missing. requiring it will break the multilib compilation of 32 bit binaries from 64 bit linux which would be a build regression. pkg-config also does not understand cross compilation unless cmake calls it with a host prefix.

can we patch libgit2's cmake to set the variables for where curl is directly instead of going through pkg-config? we're building it and in control of where it is, so we should isolate against picking up the system one unless USE_SYSTEM_CURL is asked for.

**Comment by tkelman** [comment]
If this isn't needed on windows, we just leave out the dependency. see libunwind etc. otherwise it's something that can break the build that we aren't using and don't plan to use (linking libgit2 to both it and winhttp is redundant aiui).

@vtjnash and @Keno should review the makefiles once the open items are addressed.

File: deps/libgit2.mk [hunk 6]
```diff
@@ -80,7 +80,7 @@ endif
 
 clean-libgit2:
 	-$(MAKE) -C $(BUILDDIR)/$(LIBGIT2_SRC_DIR) clean
-	-rm -f $(LIBGIT2_OBJ_TARGET)
+	-rm -f $(build_shlibdir)/libgit2* $(build_shlibdir)/libssl* $(build_shlibdir)/libcrypt
```
Comment (tkelman) [inline deps/libgit2.mk]: this is the wrong place for these

Comment (ViralBShah) [inline deps/libgit2.mk]: `make clean` should not leave things in a state that lead to an error when you do `make install` again, which is what happens without this.

**Comment by ViralBShah** [comment]
The build regression is addressed by building 32-bit linux on 32-bit linux. I don't think that needs to stop the release of this capability.

We can put an error for `pkg-config` in our Makefile, if it is not found.

**Comment by tkelman** [comment]
that makes things considerable harder to test. also not good for cross compilation which is an important capability to keep working.

File: deps/libgit2.mk (hunk 6 above)
Comment (tkelman) [inline deps/libgit2.mk]: exactly what happens? this is the wrong way to fix it

**Comment by ViralBShah** [comment]
Well, then let's find a better way to test and build without cross-compilation until we have a better solution. It is clear to me that not having proxies work for our users is strictly poorer, and we can certainly update our build infrastructure to have pkg-config installed. 

If you have a clear and simple solution, I am not opposed to it.

**Comment by tkelman** [comment]
directly set the flags that it's going through pkg-config for. we know where our source build of curl will be.

**Comment by tkelman** [comment]
have you tested this through a proxy? it can be tested on the branch until the build is working better.

**Comment by ViralBShah** [comment]
That's not how the libgit2 build works. It bails out looking for curl if it can't find `pkg-config` AFAICT. See my comments on proxy testing above.

**Comment by ViralBShah** [comment]
What do you mean by "until the build is working better"?

**Comment by tkelman** [comment]
that was on mac. what about linux? we should patch libgit2 then to not silently build without proxy support, and instead fail loudly. that's what we did for https.

File: deps/libgit2.mk (hunk 6 above)
Comment (ViralBShah) [inline deps/libgit2.mk]: It tries to copy libssl and libcrypto and fails, because they are already there.

**Comment by tkelman** [comment]
address the above todo's and issues in the makefiles

**Comment by ViralBShah** [comment]
@aviks and I have both tested on linux. I am happy to barf the whole build if `pkg-config` is not found, which will prevent silent wrong builds.

File: deps/libgit2.mk (hunk 6 above)
Comment (tkelman) [inline deps/libgit2.mk]: change it to `cp -f`? it's also not libcrypt, it's libcrypto

**Comment by ViralBShah** [comment]
@mdpradeep and I tested this with an http proxy running on a separate box from my computer, and everything working well on both mac and linux. We are testing the windows setup as well.

**Comment by mdpradeep** [comment]
Tested with [0.5.0-rc1](https://www.google.com/url?q=https%3A%2F%2Fs3.amazonaws.com%2Fjulialang%2Fbin%2Fwinnt%2Fx64%2F0.5%2Fjulia-0.5.0-rc1-win64.exe&sa=D&sntz=1&usg=AFQjCNEVdgfXosz3gtFAtUm3UKzUQQODsQ). Seems to be working fine.

The configuration required to work through proxies is as below (setting the global system level proxy does not seem to be enough !)
```
[url "https://github.com/"]
	insteadOf = git://github.com/
[https]
	proxy = http://10.128.0.4:8888
[http]
	proxy = http://10.128.0.4:8888
```

With the above config I was able to perform Package operations successfully. 

Just to verify, I changed the port to 8889 and the Package operations failed as expected.

File: deps/Makefile [hunk 7]
```diff
@@ -171,9 +171,11 @@ ifeq ($(USE_SYSTEM_LIBSSH2), 0)
 STAGE2_DEPS += libssh2
 endif
 
+ifneq ($(OS), WINNT)
```
Comment (tkelman) [inline deps/Makefile]: also needed for the makefile dependencies, this isn't enough

Comment (ViralBShah) [inline deps/Makefile]: Elaborate?

Comment (tkelman) [inline deps/Makefile]: since the makefile dependencies are not conditional on OS, this ifeq won't prevent curl from building on windows

Comment (ViralBShah) [inline deps/Makefile]: Updated libgit2.mk.

**Comment by ViralBShah** [comment]
How do we make sure travis has `pkg-config` installed? Somewhere an `apt-get` line needs to be put in for linux and the equivalent for os x (I am using brew).

File: deps/libgit2.mk [hunk 8]
```diff
@@ -45,6 +47,10 @@ ifeq ($(OS),Linux)
 $(BUILDDIR)/$(LIBGIT2_SRC_DIR)/Makefile: $(SRCDIR)/srccache/$(LIBGIT2_SRC_DIR)/libgit2-require-openssl.patch-applied $(SRCDIR)/srccache/$(LIBGIT2_SRC_DIR)/libgit2-openssl-hang.patch-applied
 endif
 $(BUILDDIR)/$(LIBGIT2_SRC_DIR)/Makefile: $(SRCDIR)/srccache/$(LIBGIT2_SRC_DIR)/CMakeLists.txt $(SRCDIR)/srccache/$(LIBGIT2_SRC_DIR)/libgit2-ssh.patch-applied $(SRCDIR)/srccache/$(LIBGIT2_SRC_DIR)/libgit2-agent-nonfatal.patch-applied $(LIBSSH2_OBJ_TARGET)
+ifneq ($(OS),WINNT)
+$(BUILDDIR)/$(LIBGIT2_SRC_DIR)/Makefile: $(CURL_OBJ_TARGET)
+endif
+$(BUILDDIR)/$(LIBGIT2_SRC_DIR)/Makefile:
```
Comment (tkelman) [inline deps/libgit2.mk]: why not leave the patches on this one

**Comment by tkelman** [comment]
It's probably already there, but see the list of brew and apt-get packages in the .travis.yml. I'm much more concerned with all the systems we don't have control over (and a few of the ones we do, like the buildbots) than I am with Travis.

File: deps/libgit2.mk (hunk 8 above)
Comment (ViralBShah) [inline deps/libgit2.mk]: I wanted to make sure that I added curl last after libssh2, which it also links against. I figured to do that, I would need to add the curl condition last.

Comment (tkelman) [inline deps/libgit2.mk]: ordering is determined by whether or not something is a dependency. if you need curl after libssh2, then libssh2 needs to be a dependency of libcurl. within the normal dependencies of a single target, the order shouldn't matter. https://www.gnu.org/software/make/manual/html_node/Prerequisite-Types.html

**Comment by ViralBShah** [comment]
I do not feel comfortable mucking with the .travis.yml. I request you put it in the right places if required. I believe this is good to go - and further improvements can be in future PRs.

**Comment by tkelman** [comment]
do you want libcurl to always be linked against libcurl, or maybe-sometimes-but-not-always depending on how parallel make happens?

> if you need curl after libssh2, then libssh2 needs to be a dependency of curl.

**Comment by tkelman** [comment]
A relatively simple thing we can do right now, to enforce that curl gets used, would be patch and change this line https://github.com/libgit2/libgit2/blob/56bbdf9349c31d6d326e4d754760549aa85bb425/CMakeLists.txt#L279 to say `REQUIRED` like we currently do for `FIND_PACKAGE(OpenSSL)`. I believe that would give an error if pkg-config is not present, and I can work on making it unnecessary in a separate PR.

**Comment by ViralBShah** [comment]
I prefer not mucking with cmake stuff. If this will get the right builds out, I prefer merging this as it is and cleaning up with your PR later.

If we use libcurl for transport, will the ssh connections also go through it, or does libgit2 handle them directly?

**Comment by tkelman** [comment]
> If we use libcurl for transport, will the ssh connections also go through it, or does libgit2 handle them directly?

I don't know the answer to that. Do ssh remotes go through a proxy? To test this, you could try the following
```
make -C deps distclean-libgit2 distclean-mbedtls distclean-libssh2 distclean-curl
make -C deps install-curl # do first so it doesn't link against ssh2, check its configure output to be sure
make -C deps install-libgit2
```

**Comment by ViralBShah** [comment]
No ssh remotes are not expected to go through proxies. I think this addresses the case at hand, of using http(s) and socks proxies. Let's get it in for the next RC.

File: deps/curl.mk (hunk 5 above)
Comment (Keno) [inline deps/curl.mk]: Yes, should be $(RPATH_ORIGIN)

File: deps/libgit2.mk (hunk 8 above)
Comment (Keno) [inline deps/libgit2.mk]: +1 to @tkelman's explanation

Comment (ViralBShah) [inline deps/libgit2.mk]: Ok, I will simplify this then a bit.

File: deps/curl.mk (hunk 5 above)
Comment (ViralBShah) [inline deps/curl.mk]: Done.

File: doc/devdocs/ast.rst [hunk 9]
```diff
@@ -267,7 +267,7 @@ Input                    AST
 ``a:b``                  ``(: a b)``
 ``a:b:c``                ``(: a b c)``
 ``a,b``                  ``(tuple a b)``
-``a==b``                 ``(comparison a == b)``
+``a==b``                 ``(call == a b)``
```
Comment (tkelman) [inline doc/devdocs/ast.rst]: messed up rebase, why is this in here?

File: README.md [hunk 10]
```diff
@@ -297,7 +299,7 @@ For a longer overview of Julia's dependencies, see these [slides](https://github
 [gcc]:          http://gcc.gnu.org
 [clang]:        http://clang.llvm.org
 [gfortran]:     https://gcc.gnu.org/fortran/
-[curl]:         http://curl.haxx.se
+[cURL]:         http://curl.haxx.se
```
Comment (tkelman) [inline README.md]: link should probably be in lower case all 3 places otherwise the "automatically download external libraries" might not link correctly

**Comment by ViralBShah** [comment]
Fixed the rpath stuff. Simplified the curl build dependency in libgit2 - and only gets done on non-windows. The Curl build now depends on mbedtls and libssh2.

Now parallel make correctly builds all the dependencies in the right after.

**Comment by ViralBShah** [comment]
The `CURL_LDFLAGS := $(RPATH_ORIGIN)` works fine on os x, but on linux, I get:
```
/usr/bin/ld: cannot find origin: No such file or directory
```

File: doc/devdocs/ast.rst (hunk 9 above)
Comment (ViralBShah) [inline doc/devdocs/ast.rst]: Crap. How do I fix this?

Comment (tkelman) [inline doc/devdocs/ast.rst]: `git rebase -i origin/master` and deselect spencer's commit, then force push over this `vs/curl` branch

**Comment by ViralBShah** [comment]
`RPATH_ESCAPED_ORIGIN` works.

**Comment by ViralBShah** [comment]
Got rid of the extra commit that had snuck in. 

Also tested successfully again on linux and os x.

**Comment by ViralBShah** [comment]
Do we have the nightlies with curl yet? We should ping all those who reported the proxy issue to test the nightlies and confirm, and that should be good enough to get it into the RC2.

**Pull request merged**